### PR TITLE
Add first implementation of did:key and Core Properties tests.

### PR DESCRIPTION
**Pull request by msporny**

### CODE REVIEW

File: packages/did-core-test-server/suites/did-spec/defaultSuiteConfig.json
Comment (msporny) [inline packages/did-core-test-server/suites/did-spec/defaultSuiteConfig.json:4]: @OR13 -- I expect we're going to have to label things like this given that there will be multiple implementations and types of did:key tests. Thoughts?

File: packages/did-core-test-server/suites/did-spec/did-core-properties.js
Comment (msporny) [inline packages/did-core-test-server/suites/did-spec/did-core-properties.js:9]: @OR13 -- Thinking of prefacing tests with the section they're in to make it easier to map these sorts of things to sections of the specification. Thoughts?

File: packages/did-core-test-server/suites/did-spec/did-core-properties.js
Comment (msporny) [inline packages/did-core-test-server/suites/did-spec/did-core-properties.js:13]: Jest's expect syntax leaves something to be desired... not a fan.

File: packages/did-core-test-server/suites/did-spec/did-core-properties.js
Comment (msporny) [inline packages/did-core-test-server/suites/did-spec/did-core-properties.js:23]: Deep nesting of tests seems to be unavoidable. :(

File: packages/did-core-test-server/suites/did-spec/utils.js
Comment (msporny) [inline packages/did-core-test-server/suites/did-spec/utils.js:21]: Disagree with uppercase "DID", and "URL" in function names. Keeping it like this to go with the established code practices in this repo.

File: packages/did-core-test-server/suites/did-spec/utils.js
Comment (msporny) [inline packages/did-core-test-server/suites/did-spec/utils.js:22]: Feels like there is going to be a ton of duplication w/ this test suite and the Transmute one: https://github.com/transmute-industries/did-core/blob/main/packages/did-core-tests/json-schemas/did.json

Are there any plans to merge these two things?

File: packages/did-core-test-server/suites/did-spec/default.json
Comment (msporny) [inline packages/did-core-test-server/suites/did-spec/default.json:5]: I expect this file to get very unwieldy -- we have multiple implementations coming down the pipeline -- can we break it up into implementations?

File: test-vectors/did-spec.latest.json
Comment (msporny) [inline test-vectors/did-spec.latest.json:49]: @OR13 is this an issue? I'd like us to group tests by section of the spec, if possible. Make it obvious which tests are testing which part of the spec. Speak clearly to which DID Methods pass/fail which specific statements and make it easy for developers to find these items.

**Comment by msporny** [comment]
Oh no, Merge is supported for this repo -- please don't do regular merges. We need to disable this -- @iherman could you please either 1) give the CODEOWNERs access to the settings for this repo... OR, disable regular merges, and make sure to only allow Squash and Rebase merges.

**Comment by iherman** [comment]
> Oh no, Merge is supported for this repo -- please don't do regular merges. We need to disable this -- @iherman could you please either 1) give the CODEOWNERs access to the settings for this repo... OR, 2) disable regular merges, and make sure to only allow Squash and Rebase merges.

I have disabled merge commits. The setup seems to be identical to the did-core settings now...

File: packages/did-core-test-server/suites/did-spec/did-core-properties.js
Comment (OR13) [reply packages/did-core-test-server/suites/did-spec/did-core-properties.js:13]: yeah that sounds good. I wrote this respec to jest code gen here: https://github.com/transmute-industries/did-core/tree/main/packages/did-core-tests/__generated__

which automatically handled all this stuff, but it harder to do that for tests related to did methods vs the spec itself.

File: packages/did-core-test-server/suites/did-spec/did-core-properties.js
Comment (OR13) [reply packages/did-core-test-server/suites/did-spec/did-core-properties.js:17]: I'm a fan! https://jestjs.io/docs/expect see also https://github.com/jest-community/jest-extended

The line you hilighted is asserting the return value of a helper function... you could rewrite the helper function to `assertValidDID` and have it throw instead of return a boolean.

File: packages/did-core-test-server/suites/did-spec/did-core-properties.js
Comment (OR13) [reply packages/did-core-test-server/suites/did-spec/did-core-properties.js:374]: yep, and problem gets worse the more we attempt to have all did methods share the same configs... lots of conditional nested code paths.

File: packages/did-core-test-server/suites/did-spec/utils.js
Comment (OR13) [reply packages/did-core-test-server/suites/did-spec/utils.js]: yeah, i think the official recomendation is to use `Did` or `Url` when you are defining a function / method... obviously the ship has sailed for `new URL("https://example.com")`

File: packages/did-core-test-server/suites/did-spec/utils.js
Comment (OR13) [reply packages/did-core-test-server/suites/did-spec/utils.js]: I can published the package and we can use it as a dependency here... but I'm not going to try to maintain our test suite in a W3C repo governed by W3C Consensus :)

File: packages/did-core-test-server/suites/did-spec/default.json
Comment (OR13) [reply packages/did-core-test-server/suites/did-spec/default.json:5]: yeah, we may want to switch to `.js` and use require to build the json instead... like we have done here:

https://github.com/w3c-ccg/vc-http-api/blob/main/packages/vc-http-api-test-server/config/Transmute.js#L43

File: test-vectors/did-spec.latest.json
Comment (OR13) [reply test-vectors/did-spec.latest.json:49]: I don't think its an issue, but like most software I expect the test suite to change as people try to actually use it.

**Review by OR13 — APPROVED** [review]

File: packages/did-core-test-server/suites/did-spec/did-core-properties.js
Comment (msporny) [reply packages/did-core-test-server/suites/did-spec/did-core-properties.js:17]: > you could rewrite the helper function to assertValidDID and have it throw instead of return a boolean.

Yes, as @shigeya raised in an issue, that should probably be what we do to make the tests easier to read. Granted, I'm not looking forward to going back and rewriting lots of tests. :)

File: packages/did-core-test-server/suites/did-spec/utils.js
Comment (msporny) [reply packages/did-core-test-server/suites/did-spec/utils.js]: > yeah, i think the official recomendation is to use Did or Url

Would you object to a global search/replace to move to that model? I'm concerned that people might look at the test suite and then duplicate the mistake in their code.

I can do the search/replace if you don't object.

File: packages/did-core-test-server/suites/did-spec/utils.js
Comment (msporny) [reply packages/did-core-test-server/suites/did-spec/utils.js]: > I'm not going to try to maintain our test suite in a W3C repo governed by W3C Consensus

What could possibly go wrong? :P 

Don't worry about it -- I expect we'll be able to get through it w/o the JSON Schema stuff... but would have probably made the tests easier/nicer than just dumping stuff in `util.js`.

File: packages/did-core-test-server/suites/did-spec/default.json
Comment (msporny) [reply packages/did-core-test-server/suites/did-spec/default.json:5]: > yeah, we may want to switch to `.js` and use require to build the json instead...

+1, let's do that before we get too many new entries and it becomes major surgery.

File: packages/did-core-test-server/suites/did-spec/defaultSuiteConfig.json
Comment (OR13) [reply packages/did-core-test-server/suites/did-spec/defaultSuiteConfig.json:4]: yeah, probably there is better ways of handling this, but for now, this is good.

File: packages/did-core-test-server/suites/did-spec/utils.js
Comment (shigeya) [reply packages/did-core-test-server/suites/did-spec/utils.js]: I'm not too fond of duplication 😝    I think having a common library in this package is a good thing. Anyone who doesn't want to depends on this package can create a similar function in their package.

File: packages/did-core-test-server/suites/did-spec/defaultSuiteConfig.json
Comment (peacekeeper) [inline packages/did-core-test-server/suites/did-spec/defaultSuiteConfig.json:41]: This is DID Resolution Metadata, not DID Document Metadata.

Comment (peacekeeper) [reply packages/did-core-test-server/suites/did-spec/defaultSuiteConfig.json:41]: Also, it should be `contentType`, not `content-type`.

**Comment by peacekeeper** [comment]
Just a small comment, I think it would be nicer if the did:key configuration is added after did:example, not inserted before it.

Also, wouldn't it be better to have implementation-specific test inputs/outputs in separate files, or do we want to have everything in a single gigantic "defaultSuiteConfig.json" of "default.json" file (and what's the difference between these two anyway?)

File: packages/did-core-test-server/suites/did-spec/defaultSuiteConfig.json
Comment (msporny) [reply packages/did-core-test-server/suites/did-spec/defaultSuiteConfig.json:41]: This fix is being dealt with in PR #37.

**Comment by msporny** [comment]
> Just a small comment, I think it would be nicer if the did:key configuration is added after did:example, not inserted before it.

Ok, I can move that.

> Also, wouldn't it be better to have implementation-specific test inputs/outputs in separate files, or do we want to have everything in a single gigantic "defaultSuiteConfig.json" of "default.json" file (and what's the difference between these two anyway?)

Agree that we want implementation-specific tests in separate files, that was discussed in https://github.com/w3c/did-test-suite/pull/38#discussion_r598914926.

@OR13, I'm also confused about the difference of `defaultSuiteConfig.json` and `default.json`... I know one of them is used to generate the report (which I have no idea how I get my new tests to show up in that file, help @OR13), and the other is used when doing the command-line tests. I'd prefer that we just had one default config.

**Comment by msporny** [comment]
> > Just a small comment, I think it would be nicer if the did:key configuration is added after did:example, not inserted before it.
> 
> Ok, I can move that.

Done in 424cce527d42ec7de01deb5ae2bc2362f83467f2.

I believe this PR is ready to merge. I'll merge it in 24 hours unless there are objections.

File: packages/did-core-test-server/suites/did-spec/did-core-properties.js
Comment (peacekeeper) [inline packages/did-core-test-server/suites/did-spec/did-core-properties.js:373]: What does this do? It feels strange that testing of Core Properties somehow depends on media types?

File: packages/did-core-test-server/suites/did-spec/did-core-properties.js
Comment (peacekeeper) [inline packages/did-core-test-server/suites/did-spec/did-core-properties.js:8]: What does this do? See comment below

**Comment by shigeya** [comment]
Is it reasonable to merge method-specific configuration?

Do we want to use a similar pattern as we separate DID Core and DID registry?

If everybody who register DID method touching configuration in this repository too, we need to accept PR to both this code and DID registry, that's not a great idea.

It may be possible to have a JSON object in DID registry, which this code (optionally) read. Then, we can reduce the number of merges.

File: packages/did-core-test-server/suites/did-spec/did-core-properties.js
Comment (OR13) [reply packages/did-core-test-server/suites/did-spec/did-core-properties.js:373]: We discussed this issue on the working group call, its an artifact of the fact that we are:

1. using json fixtures for the ADM
2. testing serialization of the ADM in JSON (did+ld+json and did+json)

I spent a good amount of time thinking about how to make this better without increasing the burden on test implementers, and this is what I came up with.

File: packages/did-core-test-server/suites/did-spec/did-core-properties.js
Comment (OR13) [reply packages/did-core-test-server/suites/did-spec/did-core-properties.js:8]: https://github.com/w3c/did-test-suite/pull/38#discussion_r601585709

**Comment by OR13** [comment]
> Is it reasonable to merge method-specific configuration?
> 

Yes, but we may refactor in the future, once more contributors have figured out how this works.

> Do we want to use a similar pattern as we separate DID Core and DID registry?
> 

Yes, imo we need to handle code splitting and json object merging from smaller files to make this easier to use... PRs welcome.

> If everybody who register DID method touching configuration in this repository too, we need to accept PR to both this code and DID registry, that's not a great idea.
> 

The did spec registries only registers did methods, so its not really a good example.... https://github.com/w3c-ccg/vc-http-api is a better example since it has apis for "issuer", "holder", "verifier"... and you may not care to test all 3, and you may not care to use features of any given one.... replace "issuer", "holder", "verifier" with 'did+json', 'did+ld+json' and 'did+cbor', and internal features such as 'revocation list' with 'canonicalId' or 'service'.

> It may be possible to have a JSON object in DID registry, which this code (optionally) read. Then, we can reduce the number of merges.

-1 to splitting test fixtures across repos until they are stable, and even then, probably not a good idea.

**Comment by shigeya** [comment]
> Yes, but we may refactor in the future, once more contributors have figured out how this works.

> Yes, imo we need to handle code splitting and json object merging from smaller files to make this easier to use... PRs welcome.

> -1 to splitting test fixtures across repos until they are stable, and even then, probably not a good idea.

As far as we plan to refactor later, I fully agree. We want to minimize dependency.

> The did spec registries only registers did methods, so its not really a good example

I understand. Thanks.

**Pull request merged**
